gdx/src/com/badlogic/gdx/utils/Queue.java

```diff
@@ -0,0 +1,181 @@
+    /** Create a new Queue which can hold `initialSize` values without resizing. */
+    public Queue(int initialSize) {
+        //noinspection unchecked
+        values = (T[]) new Object[initialSize + 1];
+    }
+
+    /** Create a new Queue which can hold `initialSize` values without resizing.
+     * This creates backing array of correct type via reflection.
+     * Use this only if you are accessing backing array directly.
+     * <p/>
+     * NOTE: Worth using only if you know what are you doing. */
+    public Queue(int initialSize, Class<T> type) {
+        //noinspection unchecked
+        values = (T[]) ArrayReflection.newInstance(type, initialSize + 1);
+    }
+
+    /** Enqueue given object.
+     * Unless backing array needs resizing, operates in O(1) time.
+     * @param object can be null
+     */
+    public void add(T object){
+        final T[] values = this.values;
+        int pushIndex = this.pushIndex;
+        final int popIndex = this.popIndex;
+
+        values[pushIndex] = object;
+        pushIndex++;
+        if(pushIndex == values.length){
+            pushIndex = 0;
+        }
+
+        if(pushIndex == popIndex){
+            //Must resize
+            final int newSize = values.length << 1;
+            //noinspection unchecked
+            final T[] newArray = (T[]) ArrayReflection.newInstance(values.getClass().getComponentType(), newSize);
+            final int currentSize = values.length;
+            for (int i = 0; i < currentSize; i++) {
+                newArray[i] = values[(popIndex + i) % currentSize];
+            }
+            this.popIndex = 0;
+            pushIndex = currentSize;
+
+            this.values = newArray;
+        }
+        this.pushIndex = pushIndex;
+    }
+
+    /** Dequeue next object in queue
+     * @return next object in queue or null if empty */
+    public T remove(){
+        return remove(null);
+    }
+
+    /** Dequeue next object in queue
+     * @return next object in queue or `defaultValue` if empty */
+    public T remove(T defaultValue){
+        final T[] values = this.values;
+        int popIndex = this.popIndex;
+
+        if(popIndex == pushIndex){
+            //Underflow
+            return defaultValue;
+        }
+        T result = values[popIndex];
+        values[popIndex] = null;
+
+        popIndex++;
+        if(popIndex == values.length){
+            popIndex = 0;
+        }
+        this.popIndex = popIndex;
+
+        return result;
+    }
+
+    /** Same as {@link Queue#remove()} but the value is kept in the queue. */
+    public T peek(){
+        return peek(null);
+    }
+
+    /** Same as {@link Queue#remove(T)} but the value is kept in the queue. */
+    public T peek(T defaultValue){
+        final int popIndex = this.popIndex;
+
+        if(popIndex == pushIndex){
+            //Underflow
+            return defaultValue;
+        }
+        return values[popIndex];
+    }
+
+    /** @return true iff this queue holds no values to remove */
```

Comment:
It stands for [if and only if](https://en.wikipedia.org/wiki/If_and_only_if). But it could be changed.